arangod/Aql/OutputAqlItemRow.cpp

```diff
@@ -0,0 +1,115 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2018 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+/// @author Michael Hackstein
+/// @author Heiko Kernbach
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#include "OutputAqlItemRow.h"
+
+#include "Aql/AqlItemBlock.h"
+#include "Aql/AqlValue.h"
+#include "Aql/ExecutorInfos.h"
+#include "Aql/InputAqlItemRow.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+OutputAqlItemRow::OutputAqlItemRow(
+    std::unique_ptr<OutputAqlItemBlockShell> blockShell)
+    : _blockShell(std::move(blockShell)),
+      _baseIndex(0),
+      _inputRowCopied(false),
+      _lastSourceRow{CreateInvalidInputRowHint{}},
+      _numValuesWritten(0) {
+  TRI_ASSERT(_blockShell != nullptr);
+}
+
+void OutputAqlItemRow::setValue(RegisterId registerId,
+                                InputAqlItemRow const& sourceRow,
+                                AqlValue const& value) {
+  if (!isOutputRegister(registerId)) {
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_WROTE_IN_WRONG_REGISTER);
+  }
+  // This is already implicitly asserted by isOutputRegister:
+  TRI_ASSERT(registerId < getNrRegisters());
+  if (_numValuesWritten >= numRegistersToWrite()) {
```

Comment:
I think we need to discuss this again.
These checks are absolutely correct and necessary, but this is nothing that is ever allowed to go wrong on a running instance, only during development.
So this if could potentially be replaced by an assertion to avoid runtime costs.